### ISSUE TITLE
Cannot migrate without any modification

### ISSUE THREAD

**Issue by shtse8** [prisma/prisma#14461]
### Bug description

> Environment variables loaded from .env
Prisma schema loaded from schema.prisma
Datasource "db": MySQL database "..." at "..."

The migration `20220721213444_increase_url_length` was modified after it was applied.
√ We need to reset the MySQL database "..." at "...".
Do you want to continue? All data will be lost. ... no

We haven't modified any migration file and it is in sync with our database.
We can't migrate our new schema. What can we do now?

### Environment & setup

- OS: Windows
- Database: Mysql
- Node.js version: 18.6

### Prisma Version

```
Environment variables loaded from .env
prisma                  : 4.1.0
@prisma/client          : 4.1.0
Current platform        : windows
Query Engine (Node-API) : libquery-engine 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 8d8414deb360336e4698a65aa45a1fbaf1ce13d8 (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 8d8414deb360336e4698a65aa45a1fbaf1ce13d8
Studio                  : 0.467.0

```

**Comment by shtse8** [comment]
ok, I can migrate now after delete the rollback record in the last migration.
It seems there is a checking issue on rollback record.

**Comment by janpio** [comment]
So the theory would be that `migrate dev` (that is the command giving you the error?) is still checking the checksum of the rolled back migration file although it is marked as "rolled back" in the database, and not the new hash from the successfully applied migration?

**Comment by shtse8** [comment]
```
```
PS D:\project> prisma migrate dev
Environment variables loaded from .env
Prisma schema loaded from schema.prisma
Datasource "db": MySQL database "project" at "127.0.0.1:3306"

Database error code: 1062

Database error:
Duplicate entry '138-5' for key 'Article.Article_articleRawId_localeId_key'

Please check the query number 1 from the migration file.


PS D:\project> prisma migrate dev
Environment variables loaded from .env
Prisma schema loaded from schema.prisma
Datasource "db": MySQL database "project" at "127.0.0.1:3306"

The migration `20220803064950_add_article_raw_id_locale_id` failed.
? We need to reset the MySQL database "project" at "127.0.0.1:3306".
Do you want to continue? All data will be lost. » (y/N)
```

What is the proper way to migrate? Everything time I failed the first time, after fixing the record and time to rerun the migrating. it will warn me all data will be lost, if I press "y" in accident, all data in production will be lost. IT'S TERRIBLE.

**Comment by floelhoeffel** [comment]
@shtse8 it sounds a bit like you are using migrate's development commands in production.

**In development**
use `migrate dev` and `db push`. The development db is considered "resettable" and "seedable" as its used to develop your migrations and to test/tweak them if needed. So if a migration fails in development, try to fix it and run `migrate dev` again (it will prompt to reset the development DB and then re-try to create/apply the migration).

**In production**
use `migrate deploy` only where Prisma will never prompt to reset the prod db. If a migration fails, use `migrate diff` to troubleshoot the failed migration and either move forward (recommended, use `migrate resolve --applied` to mark the migration to be fixed forward) or backwards (use `migrate resolve --rolled-back` to mark a migration to be rolled back)

**Comment by shtse8** [comment]
> @shtse8 it sounds a bit like you are using migrate's development commands in production.
> 
> **In development** use `migrate dev` and `db push`. The development db is considered "resettable" and "seedable" as its used to develop your migrations and to test/tweak them if needed. So if a migration fails in development, try to fix it and run `migrate dev` again (it will prompt to reset the development DB and then re-try to create/apply the migration).
> 
> **In production** use `migrate deploy` only where Prisma will never prompt to reset the prod db. If a migration fails, use `migrate diff` to troubleshoot the failed migration and either move forward (recommended, use `migrate resolve --applied` to mark the migration to be fixed forward) or backwards (use `migrate resolve --rolled-back` to mark a migration to be rolled back)

oh, thanks!!! Never knows there is a `deploy` option. let me try.

I think we can do better in the example when running prisma cli.

```
Examples

  Set up a new Prisma project
  $ prisma init

  Generate artifacts (e.g. Prisma Client)
  $ prisma generate

  Browse your data
  $ prisma studio

  Create migrations from your Prisma schema, apply them to the database, generate artifacts (e.g. Prisma Client)
  $ prisma migrate dev

  Pull the schema from an existing database, updating the Prisma schema
  $ prisma db pull

  Push the Prisma schema state to the database
  $ prisma db push
```

**Comment by tomhoule** [comment]
> I think we can do better in the example when running prisma cli.

Yes you are completely right on that. Would you like to create an issue so we can schedule work on this? I am happy to do it otherwise.

**Comment by tomhoule** [comment]
I have a PR to change how we treat edited rolled back migrations in migrate dev — I am not convinced not resetting in this case is the right behaviour, but this is what we would do if we merge that: https://github.com/prisma/prisma-engines/pull/3105

Like Frederik said, `deploy` for production databases and `migrate dev` for development databases you can reset frequently is what Migrate is designed for.

If you reallly cannot afford to reset, a combination of `migrate diff`, `db execute` and `migrate deploy` — or a different migration tool in combination with `db pull` are better choices. We are aware this is not optimal, and we will work on this at some point, but it of course competes with other work.
